From 78e4774bc5d865758b269f9de031236832ca3f18 Mon Sep 17 00:00:00 2001
From: Gregory Norton <gregory@serenade.local>
Date: Sat, 30 Aug 2026 00:00:00 +0000
Subject: [PATCH 4/7] LibCore: Allow overriding the resource root via $SERENITY_RES

On host systems /res does not exist. When nothing has installed a custom
ResourceImplementation, honor $SERENITY_RES as the resource root (e.g.
pointing at a Serenity checkout's Base/res) before falling back to /res.

diff --git a/Userland/Libraries/LibCore/ResourceImplementation.cpp b/Userland/Libraries/LibCore/ResourceImplementation.cpp
index 2834124ace..5d5f27f934 100644
--- a/Userland/Libraries/LibCore/ResourceImplementation.cpp
+++ b/Userland/Libraries/LibCore/ResourceImplementation.cpp
@@ -8,6 +8,8 @@
 #include <LibCore/ResourceImplementation.h>
 #include <LibCore/ResourceImplementationFile.h>
 #include <LibCore/System.h>
+#include <cstdlib>
+#include <cstring>
 
 namespace Core {
 
@@ -20,8 +22,14 @@ void ResourceImplementation::install(OwnPtr<ResourceImplementation> the)
 
 ResourceImplementation& ResourceImplementation::the()
 {
-    if (!s_the)
-        install(make<ResourceImplementationFile>("/res"_string));
+    if (!s_the) {
+        // On host systems (Lagom builds) /res does not exist; allow overriding
+        // the resource root, e.g. to point at a Serenity checkout's Base/res.
+        if (auto const* res_root = std::getenv("SERENITY_RES"))
+            install(make<ResourceImplementationFile>(String::from_utf8_with_replacement_character(StringView(res_root, std::strlen(res_root)))));
+        else
+            install(make<ResourceImplementationFile>("/res"_string));
+    }
     return *s_the;
 }
 
